api/apps/api/src/modules/access-control/projects-acl/project-acl.service.ts

```diff
@@ -92,4 +93,53 @@ export class ProjectAclService implements ProjectAccessControl {
       this.canViewSolutionRoles,
     );
   }
+
+  async checkUserIsOwner(userId: string, projectId: string): Promise<void> {
+    const userIsProjectOwner = await this.roles.findOne({
+      where: {
+        projectId,
+        userId,
+        roleName: Roles.project_owner,
+      },
+    });
+    if (!userIsProjectOwner) {
+      throw new ForbiddenException();
+    }
+  }
+
+  async findUsersInProject(
+    projectId: string,
+  ): Promise<UsersProjectsApiEntity[]> {
+    const usersInProject = await this.roles.find({
+      where: { projectId },
+      select: ['roleName'],
+      relations: ['user'],
+    });
+    return usersInProject;
+  }
+
+  async updateUserInProject(
+    updateUserInProjectDto: UserRoleInProjectDto,
+  ): Promise<void> {
+    const { userId, projectId, roleName } = updateUserInProjectDto;
+    const existingUserInProject = await this.roles.findOne({
+      where: {
+        projectId,
+        userId,
+      },
+    });
+    if (!existingUserInProject) {
+      await this.roles.save({
+        projectId,
+        userId,
+        roleName,
+      });
+    } else {
+      await this.roles.update({ projectId, userId }, { roleName });
+    }
+  }
```

Comment:
Correct me if I am wrong but it seems we are not checking if the request contains allowed roles: for example, if user is `viewer` / `collaborator`, can he update this rights to `owner`? (relevant spec would help us to discover it)

Reply:
good point, we'll need to check (and add tests) for this as well:
- only owners can add other owners;
- we can only remove owners if there's at least one leftover after removing (no orphan projects);

Reply:
I can't believe I overlooked this because what you said makes total sense and would be the first thing I should have realized. I added some extra tests to check that this can't be done in the e2e tests and also a condition to avoid deleting the last owner in the project.